site/content/features/sealedtypes.md

```diff
@@ -0,0 +1,506 @@
+   }
+
+   public static void main(String[] args) throws Exception {
+      walk(DirectoryEntry.of(Paths.get("/home")), e -> System.out.println(e.getPath())); 
+   }   
+}
+```
+
+Then the permitted subclasses are all direct subclasses of the sealed class in the same source file. If you want the subclasses to be public, they must be nested classes, as in the example above.
+
+If you omit the `permits` clause and there are no direct subclasses in the same source file, a compile-time error occurs.
+
+## Sealed Interfaces
+
+An interface can be sealed just like a class. It has a fixed set of permitted direct subtypes.
+
+Here is an example. An integer sequence produces one integer after another, potentially infinitely many: 
+
+```
+public sealed interface IntSeq permits FiniteSeq, IteratedSeq, GeneratedSeq {
+   int next();
+   default boolean hasNext() { return true; }
+}
+```
+
+There may be any number of ways of implementing finite sequences, and we aren't prescriptive:
+
+```
+public non-sealed interface FiniteSeq extends IntSeq {
+   int size();
+}
+```
+
+But for infinite sequences, we only support two implementations: with a generator function or an iterator function—similar to `Stream.generate` and `Stream.iterate`.
+
+![seq.png](seq.png)
+
+As you can see, with a sealed interface, the situation is a bit more complex. Its direct subtypes can be both interfaces and classes. But the rules are the same. All direct subtypes must be listed in the `permits` clause, or be in the same source file. And they must all be `final`, `sealed`, or `non-sealed`.
+
+Sandbox with complete code:
+
+```
+public sealed interface IntSeq permits FiniteSeq, IteratedSeq, GeneratedSeq {
+   int next();
+   default boolean hasNext() { return true; }
+}
+```
+
+```
+public non-sealed interface FiniteSeq extends IntSeq {
+   int size();
+}
+```
+
+```
+import java.util.function.*;
+
+public final class GeneratedSeq implements IntSeq {
+   private IntSupplier gen;
+   public GeneratedSeq(IntSupplier gen) { this.gen = gen; }
+   public int next() {
+      return gen.getAsInt();
+   }
+}
+```
+
+```
+import java.util.function.*;
+
+final class IteratedSeq implements IntSeq {
+   private IntUnaryOperator op;
+   private int value;
+   public IteratedSeq(IntUnaryOperator op, int seed) {
+      this.op = op;
+      value = seed;
+   }
+   public int next() {
+      int result = value;
+      value = op.applyAsInt(value);
+      return result;
+   }   
+}
+```
+
+```
+public class Sandbox {
+   public static int sum(IntSeq seq, int n) {
+      int sum = 0;
+      while (seq.hasNext() && n > 0) { sum += seq.next(); n--; }
+      return sum;
+   }
+   
+   public static void main(String[] args) {
+      var positiveIntegers = new IteratedSeq(x -> x + 1, 1);
+      System.out.println(sum(positiveIntegers, 100));
+   }
+}
+```
+
+# Records and Enums
+
+A sealed class can be extended by a record, which is implicitly `final`. Consider the classic example of a Lisp-style list:
```

Comment:
AFAIK records cannot extend classes. Should this read "A sealed interface can be extended by a record"

Reply:
Fixed to "A sealed interface can be implemented by a record"